appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockChartData.java

```diff
@@ -0,0 +1,405 @@
+     */
+    public void addToChart(MockChart chart) {
+        // Hide widget (MockChartData component handled by Chart)
+        iconWidget.setVisible(false);
+        iconWidget.setHeight("0");
+        iconWidget.setWidth("0");
+
+        // Set references for Chart view and Chart model
+        this.chart = chart;
+        this.chartDataModel = chart.createDataModel();
+
+        // Set the properties to the Data Series
+        setDataSeriesProperties();
+
+        // Change the visibilities of the styling properties
+        // according to Chart type
+        changeStylingPropertiesVisibility();
+
+        // Refresh the Chart view
+        refreshChart();
+    }
+
+    @Override
+    protected boolean isPropertyVisible(String propertyName) {
+        // If the set of properties to hide contains the specified
+        // property, then hide the property.
+        if (hiddenProperties.contains(propertyName)) {
+            return false;
+        }
+
+        return super.isPropertyVisible(propertyName);
+    }
+
+    @Override
+    protected void onSelectedChange(boolean selected) {
+        super.onSelectedChange(selected);
+        removeStyleDependentName("selected"); // Force remove styling
+    }
+
+    @Override
+    public void onRemoved() {
+        super.onRemoved();
+        setSourceProperty(""); // Unset the Source property to remove listener references
+        chartDataModel.removeDataSeriesFromChart(); // Remove the Data Series from the Chart
+        refreshChart();
+    }
+
+    /**
+     * Sets the current elements of the Chart Data component.
+     *
+     * @param text  new elements text
+     */
+    private void setElementsFromPairsProperty(String text){
+        currentElements = text;
+
+        chartDataModel.setElements(currentElements);
+    }
+
+    /**
+     * Sets the Source property of the Chart Data component.
+     * Responsible for showing/hiding properties according to the
+     * attached Source component type.
+     *
+     * @param source  Name of the new Source component attached
+     */
+    private void setSourceProperty(String source) {
+        // If a Data Source was previously assigned and is of type DataFile,
+        // de-attach this Data component from it
+        if (dataSource instanceof MockDataFile) {
+            ((MockDataFile)dataSource).removeDataFileChangeListener(this);
+        }
+
+        // Get the newly attached Source component
+        dataSource = editor.getComponents().getOrDefault(source, null);
+
+        changeSourcePropertiesVisibility();
+
+        // If the Data Source is now null, set back the
+        // currentElements property.
+        if (dataSource == null) {
+            onPropertyChange(PROPERTY_PAIRS, currentElements);
+        }
+
+        // If the component is currently selected, re-select it to refresh
+        // the Properties panel.
+        if (isSelected()) {
+            onSelectedChange(true);
```

Comment:
@ewpatton This is to also refresh the properties of the component, because upon changing the Source, certain properties could be hidden, and new ones are shown. Refreshing the form seems to not refresh the component panel. I have not found an alternative other than using onSelectedChange. Any ideas on which call could do it then?